clang/include/clang/Sema/Sema.h

```diff
@@ -12556,7 +12556,7 @@ class Sema final {
 private:
   // We store SYCL Kernels here and handle separately -- which is a hack.
   // FIXME: It would be best to refactor this.
-  SmallVector<Decl*, 4> SyclDeviceDecls;
+  llvm::SetVector<Decl *> SyclDeviceDecls;
```

Comment:
Why make this change?  Switching to a setVector doesn't seem to help anything here, particularly when we are using it like a vector.

Reply:
SetVector allows both deduplication and deterministic iteration order. The former is needed because SYCL_EXTERNAL functions can also be called from kernels, in which case addSyclDeviceDecl(FD) on vector would lead to duplicates. The latter is needed for deterministic compiler behavior.

Reply:
Thanks for clarifying.  Can you elaborate the problem better in the pull request message?  I still don't have a good hold of the problem you're trying to solve.

Reply:
It is all in the test case I added:

[[cl::intel_reqd_sub_group_size(2)]] void sg_size2() {}
[[cl::intel_reqd_sub_group_size(4)]] __attribute__((sycl_device)) void sg_size4() {
  sg_size2();
}

SYCL_EXTERNAL function with reqd_sub_group_size 4 calls a function with reqd_sub_group_size 2. This shoud be an error. Existing implementation detects it only when there is a kernel in place of SYCL_EXTERNAL function, and silently generates incorrect code for SYCL_EXTERNAL.

Reply:
SYCL_EXTERNAL == __attribute__((sycl_device)), BTW